File: src/test/java/no/nav/sbl/service/ContextServiceTest.java
package no.nav.sbl.service;


import no.nav.sbl.db.dao.EventDAO;
import no.nav.sbl.db.domain.PEvent;
import no.nav.sbl.rest.domain.RSContext;
import org.junit.Before;
import org.junit.Test;

import java.time.LocalDateTime;
import java.util.Optional;

import static java.time.LocalDateTime.now;
import static no.nav.sbl.db.domain.EventType.NY_AKTIV_BRUKER;
import static no.nav.sbl.service.ContextService.erFortsattAktuell;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ContextServiceTest {

    private static final String BRUKER_ID = "bruker";

    private EventDAO eventDAO;
    private ContextService contextService;

    @Before
    public void setup(){
        eventDAO = mock(EventDAO.class);
        contextService = new ContextService(eventDAO);
    }

    @Test
    public void ingen_aktiv_bruker_event(){
        har_ikke_aktiv_bruker();
    }

    @Test
    public void eventer_fra_forrige_dag_regnes_ikke_som_aktuelle() {
        PEvent event = new PEvent().created(now().minusDays(1));
        boolean result = erFortsattAktuell(event);
        assertThat(result).isFalse();
    }

    @Test
    public void eventer_fra_i_dag_regnes_som_aktuelle() {
        PEvent event = new PEvent().created(now());
        boolean result = erFortsattAktuell(event);
        assertThat(result).isTrue();
    }

    @Test
    public void aktiv_bruker_event(){
        LocalDateTime now = now();
        gitt_sist_aktive_bruker_event(now);
        assertThat(contextService.hentAktivBruker("ident")).isEqualTo(new RSContext().aktivBruker(BRUKER_ID));
    }

    @Test
    public void foreldet_aktiv_bruker_event(){
        gitt_sist_aktive_bruker_event(now().minusDays(2));
        har_ikke_aktiv_bruker();

        gitt_sist_aktive_bruker_event(now().minusDays(1).plusSeconds(1)); // i går men mindre enn en dag
        har_ikke_aktiv_bruker();
    }

    private void har_ikke_aktiv_bruker() {
        assertThat(contextService.hentAktivBruker("ident")).isEqualTo(new RSContext());
    }

    private void gitt_sist_aktive_bruker_event(LocalDateTime created) {
        PEvent pEvent = new PEvent().eventType(NY_AKTIV_BRUKER.name()).verdi(BRUKER_ID).created(created);
        when(eventDAO.sistAktiveBrukerEvent(anyString())).thenReturn(Optional.of(pEvent));
    }

}

File: src/test/java/no/nav/sbl/service/ContextServiceTest.kt
package no.nav.sbl.service

import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkObject
import no.nav.sbl.config.ApplicationCluster
import no.nav.sbl.consumers.modiacontextholder.ModiaContextHolderClient
import no.nav.sbl.domain.VeilederContext
import no.nav.sbl.domain.VeilederContextType.NY_AKTIV_BRUKER
import no.nav.sbl.redis.RedisPublisher
import no.nav.sbl.redis.VeilederContextDatabase
import no.nav.sbl.rest.model.RSContext
import no.nav.sbl.service.unleash.ToggleableFeature
import no.nav.sbl.service.unleash.ToggleableFeatureService
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.time.LocalDate
import java.time.LocalDateTime

class ContextServiceTest {
    private val brukerId = "bruker"

    private val veilederContextDatabase: VeilederContextDatabase = mockk()
    private val redisPublisher: RedisPublisher = mockk()
    private val contextHolderClient: ModiaContextHolderClient = mockk(relaxed = true)
    private val toggleableFeatureService: ToggleableFeatureService =
        mockk {
            every { isEnabled(any<ToggleableFeature>()) } returns false
        }
    private val contextService: ContextService =
        ContextService(
            veilederContextDatabase,
            listOf(redisPublisher),
            contextHolderClient,
            toggleableFeatureService,
        )

    companion object {
        @JvmStatic
        @BeforeAll
        fun beforeAll() {
            mockkObject(ApplicationCluster)
            every { ApplicationCluster.isFss() } returns true
            every { ApplicationCluster.isGcp() } returns false
        }

        @JvmStatic
        fun erFortsattAktuell(veilederContext: VeilederContext): Boolean = LocalDate.now().isEqual(veilederContext.created?.toLocalDate())
    }

    @Test
    fun ingen_aktiv_bruker_event() {
        every { contextService.hentAktivBruker(any()) } returns RSContext()
        every { veilederContextDatabase.sistAktiveBrukerEvent(any()) } returns null
        har_ikke_aktiv_bruker()
    }

    private val veilederContext =
        VeilederContext(
            veilederIdent = "veilederIdent",
            contextType = NY_AKTIV_BRUKER,
            verdi = "bruker",
        )

    @Test
    fun aktiv_bruker_event() {
        val now = LocalDateTime.now()
        gitt_sist_aktive_bruker_event(now)
        val rsContext = RSContext().apply { aktivBruker = brukerId }
        assertThat(contextService.hentAktivBruker("ident")).isEqualTo(rsContext)
    }

    private fun har_ikke_aktiv_bruker() {
        assertThat(contextService.hentAktivBruker("ident")).isEqualTo(RSContext())
    }

    private fun gitt_sist_aktive_bruker_event(created: LocalDateTime) {
        val pEvent =
            veilederContext.copy(
                contextType = NY_AKTIV_BRUKER,
                verdi = brukerId,
                created = created,
            )
        every { veilederContextDatabase.sistAktiveBrukerEvent(any<String>()) } returns pEvent
    }
}
